pkg/ipam-controller/allocator/allocator_test.go

```diff
@@ -236,6 +236,144 @@ var _ = Describe("Allocator", func() {
+		Expect(node3AllocPool.EndIP.String()).To(BeEquivalentTo("172.16.0.4"))
+	})
+
+	Context("small pools", func() {
+		It("/32 pool - can allocate if no gw", func() {
+			pool := &ipamv1alpha1.IPPool{
+				ObjectMeta: v1.ObjectMeta{Name: "small-pool"},
+				Spec: ipamv1alpha1.IPPoolSpec{
+					Subnet:           "10.10.10.10/32",
+					PerNodeBlockSize: 1,
+				},
+			}
+			a := allocator.CreatePoolAllocatorFromIPPool(ctx, pool, sets.New(testNodeName1, testNodeName2))
+			node1Alloc, err := a.AllocateFromPool(ctx, testNodeName1)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node1Alloc.StartIP.String()).To(BeEquivalentTo("10.10.10.10"))
+			Expect(node1Alloc.EndIP.String()).To(BeEquivalentTo("10.10.10.10"))
+			_, err = a.AllocateFromPool(ctx, testNodeName2)
+			Expect(errors.Is(err, allocator.ErrNoFreeRanges)).To(BeTrue())
+		})
+		It("/32 pool - can't allocate if gw set", func() {
+			pool := &ipamv1alpha1.IPPool{
+				ObjectMeta: v1.ObjectMeta{Name: "small-pool"},
+				Spec: ipamv1alpha1.IPPoolSpec{
+					Subnet:           "10.10.10.10/32",
+					PerNodeBlockSize: 1,
+					Gateway:          "10.10.10.10",
+				},
+			}
+			a := allocator.CreatePoolAllocatorFromIPPool(ctx, pool, sets.New(testNodeName1, testNodeName2))
+			_, err := a.AllocateFromPool(ctx, testNodeName1)
+			Expect(errors.Is(err, allocator.ErrNoFreeRanges)).To(BeTrue())
+		})
+		It("/31 pool - can allocate 2 ips", func() {
+			pool := &ipamv1alpha1.IPPool{
+				ObjectMeta: v1.ObjectMeta{Name: "small-pool"},
+				Spec: ipamv1alpha1.IPPoolSpec{
+					Subnet:           "10.10.10.10/31",
+					PerNodeBlockSize: 2,
+				},
+			}
+			a := allocator.CreatePoolAllocatorFromIPPool(ctx, pool, sets.New(testNodeName1, testNodeName2))
+			node1Alloc, err := a.AllocateFromPool(ctx, testNodeName1)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node1Alloc.StartIP.String()).To(BeEquivalentTo("10.10.10.10"))
+			Expect(node1Alloc.EndIP.String()).To(BeEquivalentTo("10.10.10.11"))
+			_, err = a.AllocateFromPool(ctx, testNodeName2)
+			Expect(errors.Is(err, allocator.ErrNoFreeRanges)).To(BeTrue())
+		})
+		It("/31 pool - can allocate 1 ip for 2 nodes", func() {
+			pool := &ipamv1alpha1.IPPool{
+				ObjectMeta: v1.ObjectMeta{Name: "small-pool"},
+				Spec: ipamv1alpha1.IPPoolSpec{
+					Subnet:           "10.10.10.10/31",
+					PerNodeBlockSize: 1,
+				},
+			}
+			a := allocator.CreatePoolAllocatorFromIPPool(ctx, pool, sets.New(testNodeName1, testNodeName2))
+			node1Alloc, err := a.AllocateFromPool(ctx, testNodeName1)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node1Alloc.StartIP.String()).To(BeEquivalentTo("10.10.10.10"))
+			Expect(node1Alloc.EndIP.String()).To(BeEquivalentTo("10.10.10.10"))
+			node2Alloc, err := a.AllocateFromPool(ctx, testNodeName2)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(node2Alloc.StartIP.String()).To(BeEquivalentTo("10.10.10.11"))
+			Expect(node2Alloc.EndIP.String()).To(BeEquivalentTo("10.10.10.11"))
+		})
+		It("load test", func ()  {
```

Comment:
nit: name is misleading, this test is about loading from state from status right? Also, isn't this kinda duplicate of `Load single IP range`?

Reply:
yes, they are similar, but they are targeting to test a different things. The first one validates that it is possible to load (from the CR's status) allocation with the single IP. The second test validates that loading is working as expected for /31 and /32 subnets (we have a special handling for this in the code).

I renamed this test to `/31 pool - load allocations`